fix: Stop attempting to resize ZFS in cc_growpart on Linux

TheRealFalcon opened the pull request:
## Proposed Commit Message
<!-- Include a proposed commit message because PRs are squash merged
by default.

See https://www.conventionalcommits.org/en/v1.0.0/#specification
for our commit message convention.

If the change is related to a particular cloud or particular distro,
please include the "optional scope" in the summary line. E.g.,
feat(ec2): Add support for foo to the baz

Types used by this project:
feat, fix, docs, ci, test, refactor, chore
-->
```
fix: Stop attempting to resize ZFS in cc_growpart on Linux

In e520c944, growing ZFS was added to cc_growpart to support BSD
systems. However, this left Linux in a state where it will fail to
resize and log a warning anytime it encounters ZFS. This commit
ensures that the cc_growpart resize is only attempted on BSD (or any
other system that uses `growfs`). This leaves cc_resizefs to grow the
filesystem on Linux.

This is admittedly a hack solution and long term, we should ensure
that ZFS gets grown in one module, and find a better long-term
abstraction for volume managers.


```

## Merge type

- [x] Squash merge using "Proposed Commit Message"
- [ ] Rebase and merge unique commits. Requires commit messages per-commit each referencing the pull request number (#<PR_NUM>)

TheRealFalcon commented on cloudinit/config/cc_growpart.py:
```diff
@@ -280,6 +250,36 @@ def resize(self, diskdev, partnum, partdev, fs):
         return (before, get_size(partdev, fs))
 
 
+def resizer_factory(mode: str, distro: Distro, devices: list) -> Resizer:
```
This entire function was moved as-is so to be able to reference the `Resizer` return type

holmanb reviewed (commented):
I haven't tested this, but the approach looks good for now.

I agree that it's not ideal to have resize happening in different places on different OSes. I do think that growpart is a slightly more natural fit ZFS resizing, but we can bikeshed that another day.

blackboxsw commented on cloudinit/config/cc_growpart.py:
```diff
@@ -496,7 +496,12 @@ def resize_devices(resizer, devices, distro: Distro):
+        # Ideally we would grow the FS for both OSes in the same module.
+        if fs == "zfs" and isinstance(resizer, ResizeGrowFS):
             info += _call_resizer(resizer, devent, disk, ptnum, blockdev, fs)
             return info
```
This should be a continue, not a return in case there are other devices listed in `growpart: devices: [..]`
```suggestion
            continue
```

blackboxsw reviewed (changes requested):
This looks good for the time being. It retains BSD support for growfs which is what was intended and doesn't degrade the current skip behavior on all non-BSD/non-growfs environments.

We need to change that early return to a continue. But beyond that +1.


Also we need to fix the resizer class `myresizer` in the following test to subclass from  ResizeGrowFS 
 to avoid failing the unit test now. tests/unittests/config/test_cc_growpart.py::TestResizeZFS::test_zroot[zroot/ROOT/nochange-NOCHANGE]

TheRealFalcon commented:
Addressed comment from @blackboxsw , updated tests, and manually verified that ZFS systems (both from containers and a mounted disk in a VM) result in no log warnings. Ready for review.

blackboxsw reviewed (approved):
Content and behavior look good for this PR. Thanks @TheRealFalcon
